src/Shared/FrameworkLocationHelper.cs

```diff
@@ -376,17 +376,11 @@ internal static class FrameworkLocationHelper
        };
 #endif // FEATURE_WIN32_REGISTRY
 
-        private static readonly IReadOnlyDictionary<Version, DotNetFrameworkSpec> s_dotNetFrameworkSpecDict;
-        private static readonly IReadOnlyDictionary<Version, VisualStudioSpec> s_visualStudioSpecDict;
+        private static readonly IReadOnlyDictionary<Version, DotNetFrameworkSpec> s_dotNetFrameworkSpecDict = s_dotNetFrameworkSpecs.ToDictionary(spec => spec.Version);
+        private static readonly IReadOnlyDictionary<Version, VisualStudioSpec> s_visualStudioSpecDict = s_visualStudioSpecs.ToDictionary(spec => spec.Version);
```

Comment:
This looks like expensive work. Making it run with the `beforefieldinit` semantics would be OK only if we're sure it's needed everywhere the type `FrameworkLocationHelper` is mentioned (not necessarily actually used).

Reply:
Apologies for the inconsistent use of the word "use".
- "Mention" here means the runtime sees the type because it is for example compiling a method referencing it. Making a method call in the code being compiled, accessing a field, a type check, ..., all these count.
- "Use" means an instance is actually created, a static method is called, or a static field is accessed. Actually as in at run-time.

Reply:
I can say with certainty that not all usages of `FrameworkLocationHelper` depend on `s_dotNetFrameworkSpecDict` or `s_visualStudioSpecDict`. 
Example:
https://github.com/dotnet/msbuild/blob/4bdda5bd1e70d2398badb79b92d20d27b840eaab/src/Shared/FrameworkLocationHelper.cs#L135

Does that mean `s_dotNetFrameworkSpecDict` and `s_visualStudioSpecDict` should keep their initialisation in the static constructor?

Reply:
Confirming that there are scenarios where we run static initialization but end up not needing `s_dotNetFrameworkSpecDict` or `s_visualStudioSpecDict`. Due to the non-trivial cost of initializing static fields on this class I would discourage making the change in this PR to not risk regressions. Instead, please consider switching to an explicit lazy initialization of this data.

I see the static ctor run at this callstack, for example:

```
004fe734 04dcb6b8 Microsoft.Build.Shared.FrameworkLocationHelper..cctor()
004ff0ac 04dc823e Microsoft.Build.Internal.Utilities.GetEnvironmentProperties()
004ff0f0 04dc7c64 Microsoft.Build.Evaluation.ProjectCollection.get_EnvironmentProperties()
004ff124 04dc7bd0 Microsoft.Build.Evaluation.ProjectCollection.InitializeToolsetCollection(Microsoft.Build.Evaluation.ToolsetRegistryReader, Microsoft.Build.Evaluation.ToolsetConfigurationReader)
004ff13c 0248e676 Microsoft.Build.Evaluation.ProjectCollection..ctor(System.Collections.Generic.IDictionary`2, System.Collections.Generic.IEnumerable`1, System.Collections.Generic.IEnumerable`1, Microsoft.Build.Evaluation.ToolsetDefinitionLocations, Int32, Boolean, Boolean)
004ff190 0248cb74 Microsoft.Build.CommandLine.MSBuildApp.BuildProject(System.String, System.String[], System.String, System.Collections.Generic.Dictionary`2, System.Collections.Generic.Dictionary`2, Microsoft.Build.Framework.ILogger[], Microsoft.Build.Framework.LoggerVerbosity, Microsoft.Build.CommandLine.DistributedLoggerRecord[], Boolean, System.String, Int32, Boolean, System.IO.TextWriter, System.IO.TextWriter, Boolean, System.Collections.Generic.ISet`1, System.Collections.Generic.ISet`1, Boolean, Microsoft.Build.Logging.ProfilerLogger, Boolean, Boolean, Boolean, Microsoft.Build.Graph.GraphBuildOptions, Boolean, System.String[], System.String)
004ff2e0 02481e9f Microsoft.Build.CommandLine.MSBuildApp.Execute(System.String)
004ff3a4 024810e3 Microsoft.Build.CommandLine.MSBuildApp.Main()
```

where `GetEnvironmentProperties` accesses only the `programFiles32` and `programFiles64` fields.

This is further exacerbated by compiling `FrameworkLocationHelper` into multiple assemblies so in a typical build the static initialization runs twice per process, once for the copy in `Microsoft.Build`, once for the copy in `Microsoft.Build.Utilities.Core`.

In summary, this code is in need of larger refactoring. In this PR I would err on the safer side and leave it alone.

Reply:
I've changed it to be lazily initialized.